modules/gapi/src/compiler/gcompiler.cpp

```diff
@@ -102,6 +103,73 @@ namespace
         }
         return result;
     }
+
+    // Creates ADE graph from input/output proto args
+    std::unique_ptr<ade::Graph> makeGraph(const cv::GProtoArgs &ins, const cv::GProtoArgs &outs) {
+        std::unique_ptr<ade::Graph> pG(new ade::Graph);
+        ade::Graph& g = *pG;
+
+        cv::gimpl::GModel::Graph gm(g);
+        cv::gimpl::GModel::init(gm);
+        cv::gimpl::GModelBuilder builder(g);
+        auto proto_slots = builder.put(ins, outs);
+
+        // Store Computation's protocol in metadata
+        cv::gimpl::Protocol p;
+        std::tie(p.inputs, p.outputs, p.in_nhs, p.out_nhs) = proto_slots;
+        gm.metadata().set(p);
+
+        return pG;
+    }
+
+    void checkTransformations(const cv::gapi::GKernelPackage& pkg,
+                              std::vector<std::unique_ptr<ade::Graph>>& patterns) {
+        const auto& transforms = pkg.get_transformations();
+        const auto size = transforms.size();
+        if (0u == size) return;
+        GAPI_Assert(0u == patterns.size() || size == patterns.size());
+        patterns.resize(size);
+        std::vector<std::unique_ptr<ade::Graph>> substitutes(size);
+
+        // 1. pre-generate all required graphs
+        for (auto it : ade::util::zip(ade::util::toRange(transforms),
+                                    ade::util::toRange(patterns),
+                                    ade::util::toRange(substitutes))) {
+            const auto& t = std::get<0>(it);
+            auto& p = std::get<1>(it);
+            auto& s = std::get<2>(it);
+
+            auto pattern_comp = t.pattern();
+            // NB: these patterns are re-used in apply_transformation pass
+            p = makeGraph(pattern_comp.priv().m_ins, pattern_comp.priv().m_outs);
+
+            auto substitute_comp = t.substitute();
+            s = makeGraph(substitute_comp.priv().m_ins, substitute_comp.priv().m_outs);
+        }
+
+        // FIXME: verify other types of endless loops - are there any other, though?
+
+        const auto empty = [] (const cv::gimpl::SubgraphMatch& m) {
+            return m.inputDataNodes.empty() && m.startOpNodes.empty()
+                && m.finishOpNodes.empty() && m.outputDataNodes.empty()
+                && m.inputTestDataNodes.empty() && m.outputTestDataNodes.empty();
+        };
+        // 2. verify there are no patterns in substitutes
+        for (size_t i = 0; i < size; ++i) {
+            auto& p = patterns[i];
+            for (size_t j = i; j < size; ++j) {
+                auto& s = substitutes[j];
```

Comment:
hmm, your argument is valid. I remember having similar idea but then decided not to change it. I wonder why... okay, if I won't come up with a decent answer to you and myself, let's check for `j = 0` && `j != i`

Reply:
changed this part but I feel like this logic requires much more attention. @AsyaPronina please take a look

Reply:
Hi, @andrey-golubev , @dmatveev!
I couldn't understand: why we prohibit one pattern to be in a substitute  of another pattern if this doesn't produce recursion?
The previous logic just allows to pattern from the right be in substitutes on the left, which doesn't produce recursion but, unfortunately, this logic also reduces some  possible cases where recursion couldn't happen and pattern from the left may be in substitute form the right.
 
How can we handle this case?

My suggestion is the following:
If we rely on the previous logic - it will allow all cases with pattern from right to be in substitute from left to be happened w/o recursion.
To handle case when pattern from the left is in substitute from the right we shall try to swap these two transformations to fit in our logic. If we can't swap that means that existence of left pattern in right substitute leads to recursion and we shall fail our check.

Reply:
Hi, @andrey-golubev , @dmatveev!
I couldn't understand: why we prohibit one pattern to be in a substitute  of another pattern if this doesn't produce recursion?
The previous logic just allows to pattern from the right be in substitutes on the left, which doesn't produce recursion but, unfortunately, this logic also reduces some  possible cases where recursion couldn't happen and pattern from the left may be in substitute from the right.
 
How can we handle this case?

My suggestion is the following:
If we rely on the previous logic - it will allow all cases with pattern from right to be in substitute from left to be happened w/o recursion.
To handle case when pattern from the left is in substitute from the right we shall try to swap these two transformations to fit in our logic. If we can't swap that means that existence of left pattern in right substitute leads to recursion and we shall fail our check.